jumanji/environments/packing/flat_pack/env.py

```diff
@@ -0,0 +1,530 @@
+        """
+
+        default_generator = RandomFlatPackGenerator(
+            num_row_blocks=5,
+            num_col_blocks=5,
+        )
+
+        self.generator = generator or default_generator
+        self.num_row_blocks = self.generator.num_row_blocks
+        self.num_col_blocks = self.generator.num_col_blocks
+        self.num_blocks = self.num_row_blocks * self.num_col_blocks
+        self.num_rows, self.num_cols = (
+            compute_grid_dim(self.num_row_blocks),
+            compute_grid_dim(self.num_col_blocks),
+        )
+        self.reward_fn = reward_fn or CellDenseReward()
+        self.viewer = viewer or FlatPackViewer(
+            "FlatPack", self.num_blocks, render_mode="human"
+        )
+
+    def __repr__(self) -> str:
+        return (
+            f"FlatPack environment with a grid size of ({self.num_rows}x{self.num_cols}) "
+            f"with {self.num_row_blocks} row blocks, {self.num_col_blocks} column "
+            f"blocks. Each block has dimension (3x3)."
+        )
+
+    def reset(
+        self,
+        key: chex.PRNGKey,
+    ) -> Tuple[State, TimeStep[Observation]]:
+
+        """Resets the environment.
+
+        Args:
+            key: PRNG key for generating a new instance.
+
+        Returns:
+            a tuple of the initial environment state and a time step.
+        """
+
+        grid_state = self.generator(key)
+
+        obs = self._observation_from_state(grid_state)
+        timestep = restart(observation=obs)
+
+        return grid_state, timestep
+
+    def step(
+        self, state: State, action: chex.Array
+    ) -> Tuple[State, TimeStep[Observation]]:
+        """Steps the environment.
+
+        Args:
+            state: current state of the environment.
+            action: action to take.
+
+        Returns:
+            a tuple of the next environment state and a time step.
+        """
+
+        # Unpack and use actions
+        block_idx, rotation, row_idx, col_idx = action
+
+        chosen_block = state.blocks[block_idx]
+
+        # Rotate chosen block
+        chosen_block = rotate_block(chosen_block, rotation)
+
+        grid_block = self._expand_block_to_grid(chosen_block, row_idx, col_idx)
+        grid_mask_block = self._get_ones_like_expanded_block(grid_block=grid_block)
+
+        action_is_legal = self._is_legal_action(
+            action, state.current_grid, state.placed_blocks, grid_mask_block
+        )
+
+        # If the action is legal create a new grid and update the placed blocks array
+        new_grid = jax.lax.cond(
+            action_is_legal,
+            lambda: state.current_grid + grid_block,
+            lambda: state.current_grid,
+        )
+        placed_blocks = jax.lax.cond(
+            action_is_legal,
+            lambda: state.placed_blocks.at[block_idx].set(True),
+            lambda: state.placed_blocks,
+        )
+
+        new_action_mask = self._make_action_mask(new_grid, state.blocks, placed_blocks)
```

Comment:
The action mask is include in the state of the `maze`, `cleaner` and `job_shop` environments and this is where I got the idea from. My suggestion is to leave it as is.

Reply:
I guess the action mask is needed to create the observation. But also, in the environments you mentioned above, the action mask is usually used to directly compute/read the value of `action_is_legal`. If possible, could we do that as well instead of recomputing the method `self._is_legal_action`?

Reply:
This is quite a nice suggestion, I think we can do something like: 
```
action_is_legal = state.action_mask[block_idx, rotation, row_idx, col_idx]
```
Since the state will carry the action mask from the previous step. Let me see.

Reply:
Indeed, this works perfectly :)